activesupport/lib/active_support/core_ext/object/json.rb

```diff
@@ -59,6 +59,12 @@ class Object
   def as_json(options = nil) # :nodoc:
     if respond_to?(:to_hash)
       to_hash.as_json(options)
+    elsif respond_to?(:to_ary)
+      to_ary.as_json(options)
+    elsif respond_to?(:to_str)
+      to_str.as_json(options)
+    elsif respond_to?(:to_int)
+      to_int.as_json(options)
```

Comment:
All these extra checks are quire costly. Would be worth benching this with benchmark/ips to see what's the overhead is like.